File: swift_module/CommonDI/HealthCheckManagerImpl.swift
import app
import MyLibrary

public class HealthCheckManagerImpl: HealthCheckManager {
    private let configsRepository = DobbyConfigsRepositoryImpl.shared

    public func getConnectionState() -> VpnConnectionState {
        // Persisted health is intentionally not consulted here: it can outlive the extension
        // process and must never make the UI claim that a disconnected NE tunnel is connected.
        return IOSVpnConnectionAuthority.connectionState()
    }

    public func doInitHealthCheck() {
        configsRepository.setHealthCheckState(state: 0)
        configsRepository.setHealthCheckGeneration(IOSVpnConnectionAuthority.currentGeneration())
    }

    public func startHealthCheck() {
        configsRepository.setHealthCheckState(state: 1)
        configsRepository.setHealthCheckGeneration(IOSVpnConnectionAuthority.currentGeneration())
    }

    public func stopHealthCheck() {
        configsRepository.setHealthCheckState(state: 0)
        configsRepository.setHealthCheckGeneration(IOSVpnConnectionAuthority.currentGeneration())
    }

    public func measureTunnelProbeAverageLatencyMillis(timeoutMillis: Int64) -> Int64 {
        return Cloak_outlineMeasureTunnelProbeAverageLatencyMillisWithTimeout(timeoutMillis)
    }
}

File: kmp_module/app/src/androidMain/kotlin/com/dobby/feature/diagnostic/domain/HealthCheckManagerImpl.kt
package com.dobby.feature.diagnostic.domain

import com.dobby.feature.logging.Logger
import com.dobby.backend.GoBackendWrapper

class HealthCheckManagerImpl(
    private val logger: Logger,
) : HealthCheckManager {
    override fun getConnectionState(): VpnConnectionState =
        when (GoBackendWrapper.getConnectionState()) {
            0 -> VpnConnectionState.DISCONNECTED
            1 -> VpnConnectionState.CONNECTING
            2 -> VpnConnectionState.CONNECTED
            else -> {
                logger.log("[WARN] Got invalid connection state")
                VpnConnectionState.DISCONNECTED
            }
        }

    override fun initHealthCheck() {
        GoBackendWrapper.initHealthCheck()
    }

    override fun startHealthCheck() {
        GoBackendWrapper.startHealthCheck()
    }

    override fun stopHealthCheck() {
        GoBackendWrapper.stopHealthCheck()
    }

    override fun measureTunnelProbeAverageLatencyMillis(timeoutMillis: Long): Long {
        return GoBackendWrapper.measureTunnelProbeAverageLatencyMillis(timeoutMillis)
    }
}
